### PR TITLE
Add trades and offers

### PR DESCRIPTION
**Pull request by AEgan**
These changes will add the trade and offer models to the application. This only modifies the database, but will allow continued work on the trade and trade offer process in the web application. People looking to trade in the game can only post one player, and this is reflected on the application. Other users can offer multiple players for a trade, and this is reflected by the offer and players `has_and_belongs_to_many` relationship, and the join table between the two.

The trade's `partner_id` is to reflect the user who submitted an accepted offer, and by using the `partner_id` and `trade_id` we can find the accepted offer.

I started working on this branch before the last pull request was merged, so there are a few changes that were made in #2 that are not yet reflected here.

The first new commit that is not on master is 305eb3c

@metaskills @ryanbillings @nhessler

### CODE REVIEW

File: app/models/trade.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  # for people looking for trades, but are intended for people involved with the trade
+  validates_inclusion_of :status, in: ["Open", "Partner Found", "Awaiting Ratings", "Complete", "Closed"]
+
+  # scopes
+  scope :open, -> { where(status: "Open") }
+  scope :complete, -> { where(status: "Complete") }
+
+  # gets the offer that has been accepted by using the partner_id and the trade's id
+  def offer
+    Offer.where(user_id: self.partner_id, trade_id: self.id).first
```
Comment (ryanbillings) [inline app/models/trade.rb]: This is just a stylistic thing, but generally we avoid using self unless you need to (you could omit it here)

Comment (AEgan) [inline app/models/trade.rb]: Thanks for letting me know, I'll keep that in mind when working on projects here to be consistent. I'll remove it where I can in this project as well.

File: app/models/trade.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  private
+  def partner_and_offer_are_valid?
+    if User.find_by_id(self.partner_id).nil? || self.offer.nil?
+      errors.add(:base, "this isn't gunna work")
+      return false
```
Comment (ryanbillings) [inline app/models/trade.rb]: I could be wrong here, but I believe that in a custom validation, if you populate the errors hash the record will be marked as invalid, which means you wouldn't need to return false.

File: app/models/trade.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  validates_numericality_of :partner_rating, only_integer: true, allow_nil: true
+  validate -> { reference_exists_in_system("User") }
+  validate -> { reference_exists_in_system("Player") }
+  validate :partner_and_offer_are_valid?, if: :partner_id_changed?
+
+  # we could get pretty detailed with different statuses, but I figured we need
+  # a way to determine trades without a confirmed partner, completed trades to
+  # use to see ratings, and closed (canceled). The other two aren't as important
+  # for people looking for trades, but are intended for people involved with the trade
+  validates_inclusion_of :status, in: ["Open", "Partner Found", "Awaiting Ratings", "Complete", "Closed"]
```
Comment (ryanbillings) [inline app/models/trade.rb]: Maybe you could turn your statuses into constants or put them into a yaml file? It looks like some of them (like 'Open' as the default) are used in several places.

File: test/models/trade_test.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      deny no_user.valid?
+    end
+
+    should "Not allow a trade to be created without a valid player" do
+      no_player = FactoryGirl.build(:trade, user: @alex, player_id: -1, partner: @john)
+      deny no_player.valid?
+    end
+
+    should "have a method to get the offer that has been accepted for the trade" do
+      create_offers
```
Comment (ryanbillings) [inline test/models/trade_test.rb]: Nice tests! It looks like a few of these tests start with create_offers and end with destroy_offers. Would it make sense to group these tests in a nested context, and add those methods to the nested setup/teardown?

File: app/models/concerns/referenceable.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+module Referenceable extend ActiveSupport::Concern
```
Comment (ryanbillings) [inline app/models/concerns/referenceable.rb:1]: Nice on adding this concern! This is some cool stuff.

**Comment by ryanbillings** [comment]
:+1: Great work on this PR @AEgan !

File: app/models/trade.rb (hunk 2 above)
Comment (AEgan) [inline app/models/trade.rb]: Ken pointed this out as well, thanks for the heads up!

File: test/models/trade_test.rb (hunk 4 above)
Comment (AEgan) [inline test/models/trade_test.rb]: Good idea, thanks for the suggestion

File: app/models/trade.rb (hunk 3 above)
Comment (AEgan) [inline app/models/trade.rb]: Good suggestion. Is this more along the lines of what you're looking for with a yaml file?
in config/trade_statuses.yml have

```
open: "Open"
partnered: "Partner Found"
ratings: "Awaiting Ratings"
complete: "Complete"
closed: "Closed"
```

and use them like this in the Trade model

```
STATUSES = YAML.load_file("#{Rails.root}/config/trade_statuses.yml")
validates_inclusion_of :status, in: STATUSES.values

# scopes
scope :open, -> { where(status: STATUSES["open"]) }
scope :complete, -> { where(status: STATUSES["complete"]) }
```

andreference them anywhere outside of the model with `Trade::STATUSES["open"]` ?

Comment (ryanbillings) [inline app/models/trade.rb]: Nice! That looks great to me - you could probably even update [this](https://github.com/AEgan/HUTrader/blob/trading/db/migrate/20150724194534_create_trades.rb#L9) like you suggested above.

**Pull request merged**